controllers/ingressnodefirewall_controller_rules_test.go

```diff
@@ -0,0 +1,753 @@
+		Expect(k8sClient.DeleteAllOf(context.Background(), &v1.Node{})).Should(Succeed())
+		Eventually(func() bool {
+			nodeStateList := &infv1alpha1.IngressNodeFirewallNodeStateList{}
+			err := k8sClient.List(ctx, nodeStateList)
+			if err != nil {
+				fmt.Fprintf(GinkgoWriter, "Could not list IngressNodeFirewallNodeStates on cleanup, err: %q", err)
+				return false
+			}
+			return len(nodeStateList.Items) == 0
+		}).Should(BeTrue())
+	})
+
+	tcs := []struct {
+		inSpecs     []infv1alpha1.IngressNodeFirewallSpec
+		outSpec     infv1alpha1.IngressNodeFirewallNodeStateSpec
+		statusError string
+	}{
+		// TC0 - baseline test without merging.
+		{
+			inSpecs: []infv1alpha1.IngressNodeFirewallSpec{
+				{
+					Ingress: []infv1alpha1.IngressNodeFirewallRules{
+						{
+							SourceCIDRs: []string{"10.0.0.0"},
+							FirewallProtocolRules: []infv1alpha1.IngressNodeFirewallProtocolRule{
+								{
+									Order: 10,
+									ProtocolRule: &infv1alpha1.IngressNodeFirewallProtoRule{
+										Ports: "80",
+									},
+									Protocol: "tcp",
+									Action:   "allow",
+								},
+							},
+						},
+					},
+					Interfaces: &[]string{"eth0"},
+				},
+			},
+			outSpec: infv1alpha1.IngressNodeFirewallNodeStateSpec{
+				InterfaceIngressRules: map[string][]infv1alpha1.IngressNodeFirewallRules{
+					"eth0": {
+						{
+							SourceCIDRs: []string{"10.0.0.0"},
+							FirewallProtocolRules: []infv1alpha1.IngressNodeFirewallProtocolRule{
+								{
+									Order: 10,
+									ProtocolRule: &infv1alpha1.IngressNodeFirewallProtoRule{
+										Ports: "80",
+									},
+									Protocol: "tcp",
+									Action:   "allow",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		// TC1 - duplicate rules and interfaces shall throw an error due to duplicate order.
+		{
+			inSpecs: []infv1alpha1.IngressNodeFirewallSpec{
+				{
+					Ingress: []infv1alpha1.IngressNodeFirewallRules{
+						{
+							SourceCIDRs: []string{"10.0.0.0"},
+							FirewallProtocolRules: []infv1alpha1.IngressNodeFirewallProtocolRule{
+								{
+									Order: 10,
+									ProtocolRule: &infv1alpha1.IngressNodeFirewallProtoRule{
+										Ports: "80",
+									},
+									Protocol: "tcp",
+									Action:   "allow",
+								},
+							},
+						},
+					},
+					Interfaces: &[]string{"eth0"},
+				},
+				{
+					Ingress: []infv1alpha1.IngressNodeFirewallRules{
+						{
+							SourceCIDRs: []string{"10.0.0.0"},
+							FirewallProtocolRules: []infv1alpha1.IngressNodeFirewallProtocolRule{
+								{
+									Order: 10, // should throw error
+									ProtocolRule: &infv1alpha1.IngressNodeFirewallProtoRule{
+										Ports: "80",
+									},
+									Protocol: "tcp",
+									Action:   "allow",
+								},
+							},
+						},
+					},
+					Interfaces: &[]string{"eth0"},
+				},
+			},
+			outSpec:     infv1alpha1.IngressNodeFirewallNodeStateSpec{},
+			statusError: "duplicate order 10 detected",
```

Comment:
that is kind of dup to what webhook do i won't hurt to have more than safety gates but just wanted to highlight that

Reply:
I'm totally aware of that. I'm just stubborn and would like to have some validation further down, too :-)